examples/Imperative.hs

```diff
@@ -73,23 +76,37 @@ data Prog operand a
 -- thing. We want to be able to replace those variables with operand values, and
 -- that would not be possible if variables were allowed to appear inside Prog
 -- but outside of an operand.
-pInstantiate1 :: (Applicative operand, Monad operand)
+pInstantiate1 :: forall operand b a. (Applicative operand, Monad operand)
               => operand a
-              -> Prog (Scope () operand) a
+              -> Prog (Scope b operand) a
               -> Prog operand a
-pInstantiate1 x (Ret o)        = Ret (instantiate1 x o)
-pInstantiate1 x (Add o1 o2 cc) = Add (instantiate1 x o1)
-                                     (instantiate1 x o2)
-                                     (pInstantiate1 (lift x) cc)
+pInstantiate1 = go instantiate1
+  where
+    go :: forall o o' u
+        . (forall v. operand v -> o v -> o' v)
+       -> operand u -> Prog o u -> Prog o' u
+    go f x (Ret o)        = Ret (f x o)
+    go f x (Add o1 o2 cc) = Add (f x o1) (f x o2)
+                          $ go f' x cc
+      where
+        f' :: operand v -> Scope () o v -> Scope () o' v
```

Comment:
Could you comment how `f'` is different from `f`, and why we need one? Now this looks complex for no reason (maybe more obvious when you see the diff, but not so otherwise).